lightning/src/routing/router.rs

```diff
@@ -333,84 +671,298 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+		let mut found_new_path = false;
+
+		// Step (2).
+		// If this loop terminates due the exhaustion of targets, two situations are possible:
+		// - not enough outgoing liquidity:
+		//   0 < already_collected_value_msat < final_value_msat
+		// - enough outgoing liquidity:
+		//   final_value_msat <= already_collected_value_msat < recommended_value_msat
+		// Both these cases (and other cases except reaching recommended_value_msat) mean that
+		// paths_collection will be stopped because found_new_path==false.
+		// This is not necessarily a routing failure.
+		'path_construction: while let Some(RouteGraphNode { pubkey, lowest_fee_to_node, value_contribution_msat, .. }) = targets.pop() {
+
+			// Since we're going payee-to-payer, hitting our node as a target means we should stop
+			// traversing the graph and arrange the path out of what we found.
+			if pubkey == *our_node_id {
+				let mut new_entry = dist.remove(&our_node_id).unwrap();
+				let mut ordered_hops = vec!(new_entry.clone());
+
+				'path_walk: loop {
+					if let Some(&(_, ref features, _)) = first_hop_targets.get(&ordered_hops.last().unwrap().route_hop.pubkey) {
+						ordered_hops.last_mut().unwrap().route_hop.node_features = features.to_context();
+					} else if let Some(node) = network.get_nodes().get(&ordered_hops.last().unwrap().route_hop.pubkey) {
+						if let Some(node_info) = node.announcement_info.as_ref() {
+							ordered_hops.last_mut().unwrap().route_hop.node_features = node_info.features.clone();
+						} else {
+							ordered_hops.last_mut().unwrap().route_hop.node_features = NodeFeatures::empty();
+						}
 					} else {
-						res.last_mut().unwrap().node_features = NodeFeatures::empty();
+						// We should be able to fill in features for everything except the last
+						// hop, if the last hop was provided via a BOLT 11 invoice (though we
+						// should be able to extend it further as BOLT 11 does have feature
+						// flags for the last hop node itself).
+						assert!(ordered_hops.last().unwrap().route_hop.pubkey == *payee);
 					}
-				} else {
-					// We should be able to fill in features for everything except the last
-					// hop, if the last hop was provided via a BOLT 11 invoice (though we
-					// should be able to extend it further as BOLT 11 does have feature
-					// flags for the last hop node itself).
-					assert!(res.last().unwrap().pubkey == *target);
+
+					// Means we succesfully traversed from the payer to the payee, now
+					// save this path for the payment route. Also, update the liquidity
+					// remaining on the used hops, so that we take them into account
+					// while looking for more paths.
+					if ordered_hops.last().unwrap().route_hop.pubkey == *payee {
+						break 'path_walk;
+					}
+
+					new_entry = match dist.remove(&ordered_hops.last().unwrap().route_hop.pubkey) {
+						Some(payment_hop) => payment_hop,
+						// We can't arrive at None because of how we traverse with add_entry.
```

Comment:
Nit: maybe say "because, if we ever add an entry to targets, we also fill in the entry in dist" to be more specific.